pkg/model/project.go

```diff
@@ -243,3 +350,73 @@ func (p *ProjectSSOConfig_GitHub) GenerateAuthCodeURL(project, callbackURL, stat
 
 	return authURL, nil
 }
+
+// UpdateRBACRoles update only custom rbac roles.
+func (p *Project) UpdateRBACRoles(roles []*ProjectRBACRole) error {
+	v := make(map[string]struct{}, len(roles))
+	update := make([]*ProjectRBACRole, 0, len(roles))
+	for _, role := range roles {
+		if role.IsBuiltin {
+			continue
+		}
+		if role.IsBuiltinName() {
+			return fmt.Errorf("cannot use built-in role name")
+		}
+		if _, ok := v[role.Name]; ok {
+			return fmt.Errorf("role name must be unique")
+		}
+		v[role.Name] = struct{}{}
+		update = append(update, role)
+	}
+	p.RbacRoles = update
+	return nil
+}
+
+func (p *ProjectRBACRole) IsBuiltinName() bool {
+	return p.Name == builtinRBACRoleAdmin.String() ||
+		p.Name == builtinRBACRoleEditor.String() ||
+		p.Name == builtinRBACRoleViewer.String()
+}
+
+// UpdateUserGroups update user groups.
+func (p *Project) UpdateUserGroups(groups []*ProjectUserGroup) error {
+	v := make(map[string]struct{}, len(groups))
+	for _, group := range groups {
+		if _, ok := v[group.SsoGroup]; ok {
+			return fmt.Errorf("cannot assign multi roles to sso group")
+		}
+		v[group.SsoGroup] = struct{}{}
+	}
```

Comment:
How about moving this to a new `ValidateUserGroups` function?
After that, we can assign the validated data to `p.UserGroups` directly instead of using this `UpdateUserGroups` function.